# Generate fasta files for phylogenetic analysis
# Objective:
# Take the DNA Barcoding data and output it as a FASTA file that we can load into Mega
# From there, ALIGN the sequences, then pass on to MrBayes for analysis

# Todo:
# - Sort out issue with seq.name
#  . Use SampleID by itself as that can be filled in later with more information
# - Sort out the issue with the sequences
#  . Alignment is fine, but empty spaces and other junk needs to be trimmed out
#  . When run in MrBayes, we want the largest usable region.
#  . Leading and trailing whitespace needs to be deleted.

# Load libraries
library(tidyverse)
library(phylotools)
library(lubridate)

# Load all the data
kitikmeotdata <- read_tsv("data/workingdata_2024_02_05.tsv")
# Load just Cambridge Bay, Kugluktuk, Gjoa Haven, Kugaaruk, and Taloyoak
cambridgebay <- read_tsv("data/cambridgebay_2024_01_22.tsv")
kugluktuk <- read_tsv("data/kugluktuk_2024_01_22.tsv")
gjoahaven <- read_tsv("data/gjoahaven_2024_01_22.tsv")
kugaaruk <- read_tsv("data/kugaaruk_2024_01_22.tsv")
taloyoak <- read_tsv("data/taloyoak_2024_01_22.tsv")
# Clepsis
clepsis <- read_tsv("data/Clepsis_moeschleriana_BOLD-AAB4040.tsv")

# Generate fasta files for Mega ----
# Before we go any further, let's do a little filtering to generate some phylogenetic tree's in Mega
# Load phylotools so we can export as a Fasta format file
# We'll need to output two things.
# Fasta needs seq.name and seq.text
# seq.name is how it'll be referred to in the NJT generated by Mega
# The command to save as a fasta format file with Phylotools is:
# dat2fasta(dat, outfile = "out.fasta")
# Currently, there is an issue with character limits. So that will need to be resolved.

# Right, let's take all Lepidoptera from Cambridge Bay and generate a NJT phylogenetic tree
cbay_lepidoptera <- cambridgebay %>%
  filter(Order == "Lepidoptera")
cbay_lepidoptera <- cbay_lepidoptera %>%
  select(seq.data,Sector,taxon,"Sample ID")
names(cbay_lepidoptera) <- c("seq.data","Community","Taxon","SampleID")

# Next, we'll need to re-combine these to form the 'seq.name' and 'seq.text' columns needed for the fasta format
cbay_lepidoptera <- cbay_lepidoptera %>%
  mutate(
    seq.name = SampleID,
    seq.text = seq.data
    ) %>%
  select(
    seq.name,
    seq.text
  )

# Attempt to save as a fasta format file to run through Mega for alignment and tree generation
dat2fasta(cbay_lepidoptera, outfile = "data/MEGA/cbay_lepidoptera.fas")
# Do some garbage collection
rm(cbay_lepidoptera)

# We will need to repeat the same process for the other communities.
# Before we do that, let's isolate the 47 clepsis specimens from BOLD
clepsis_run <- clepsis %>%
  filter(markercode == "COI-5P") %>%
  select(nucleotides,province_state,species_name,sampleid)
names(clepsis_run) <- c("seq.data","Province","Taxon","SampleID")
# Prepare for run
clepsis_run <- clepsis_run %>%
  mutate(
    seq.name = SampleID,
    seq.text = seq.data
  ) %>%
  select(
    seq.name,
    seq.text
  )
# Save in a format that can be opened in MEGA
dat2fasta(clepsis_run, outfile = "data/MEGA/clepsis.fas")
# Do some garbage collection
rm(clepsis,clepsis_run)

# Right, so we've established that we can take these sequences and export them to be aligned in Mega before being run in MrBayes.
# Let's focus on some important questions instead.
# What questions? Well, to recap:
# 1. We have narrowed our DNA Barcodes down to terrestrial Arthropoda
# 2. We've eliminated any specimens that were not collected in our target communities in the Kitikmeot
# 3. 

# Let's do something interesting. We want to find BINs that persist year after year
cbay_2018 <- cambridgebay %>%
  subset(CollectionDate > "2018-01-01" & CollectionDate < "2018-12-31")
cbay_2019 <- cambridgebay %>%
  subset(CollectionDate > "2019-01-01" & CollectionDate < "2019-12-31")
cbay_2020 <- cambridgebay %>%
  subset(CollectionDate > "2020-01-01" & CollectionDate < "2020-12-31")
cbay_2021 <- cambridgebay %>%
  subset(CollectionDate > "2021-01-01" & CollectionDate < "2021-12-31")
cbay_2022 <- cambridgebay %>%
  subset(CollectionDate > "2022-01-01" & CollectionDate < "2022-12-31")
cbay_2023 <- cambridgebay %>%
  subset(CollectionDate > "2023-01-01" & CollectionDate < "2023-12-31")
# Do the same for Kugluktuk
kugl_2018 <- kugluktuk %>%
  subset(CollectionDate > "2018-01-01" & CollectionDate < "2018-12-31")
kugl_2019 <- kugluktuk %>%
  subset(CollectionDate > "2019-01-01" & CollectionDate < "2019-12-31")
kugl_2020 <- kugluktuk %>%
  subset(CollectionDate > "2020-01-01" & CollectionDate < "2020-12-31")
kugl_2021 <- kugluktuk %>%
  subset(CollectionDate > "2021-01-01" & CollectionDate < "2021-12-31")
kugl_2022 <- kugluktuk %>%
  subset(CollectionDate > "2022-01-01" & CollectionDate < "2022-12-31")
kugl_2023 <- kugluktuk %>%
  subset(CollectionDate > "2023-01-01" & CollectionDate < "2023-12-31")
# Do the same for Gjoa Haven
gjoa_2018 <- gjoahaven %>%
  subset(CollectionDate > "2018-01-01" & CollectionDate < "2018-12-31")
gjoa_2019 <- gjoahaven %>%
  subset(CollectionDate > "2019-01-01" & CollectionDate < "2019-12-31")
gjoa_2020 <- gjoahaven %>%
  subset(CollectionDate > "2020-01-01" & CollectionDate < "2020-12-31")
gjoa_2021 <- gjoahaven %>%
  subset(CollectionDate > "2021-01-01" & CollectionDate < "2021-12-31")
gjoa_2022 <- gjoahaven %>%
  subset(CollectionDate > "2022-01-01" & CollectionDate < "2022-12-31")
gjoa_2023 <- gjoahaven %>%
  subset(CollectionDate > "2023-01-01" & CollectionDate < "2023-12-31")
# Do the same for Kugaaruk
kuga_2018 <- kugaaruk %>%
  subset(CollectionDate > "2018-01-01" & CollectionDate < "2018-12-31")
kuga_2019 <- kugaaruk %>%
  subset(CollectionDate > "2019-01-01" & CollectionDate < "2019-12-31")
kuga_2020 <- kugaaruk %>%
  subset(CollectionDate > "2020-01-01" & CollectionDate < "2020-12-31")
kuga_2021 <- kugaaruk %>%
  subset(CollectionDate > "2021-01-01" & CollectionDate < "2021-12-31")
kuga_2022 <- kugaaruk %>%
  subset(CollectionDate > "2022-01-01" & CollectionDate < "2022-12-31")
kuga_2023 <- kugaaruk %>%
  subset(CollectionDate > "2023-01-01" & CollectionDate < "2023-12-31")
# Do the same for Taloyoak
taloyoak_2018 <- taloyoak %>%
  subset(CollectionDate > "2018-01-01" & CollectionDate < "2018-12-31")
taloyoak_2019 <- taloyoak %>%
  subset(CollectionDate > "2019-01-01" & CollectionDate < "2019-12-31")
taloyoak_2020 <- taloyoak %>%
  subset(CollectionDate > "2020-01-01" & CollectionDate < "2020-12-31")
taloyoak_2021 <- taloyoak %>%
  subset(CollectionDate > "2021-01-01" & CollectionDate < "2021-12-31")
taloyoak_2022 <- taloyoak %>%
  subset(CollectionDate > "2022-01-01" & CollectionDate < "2022-12-31")
taloyoak_2023 <- taloyoak %>%
  subset(CollectionDate > "2023-01-01" & CollectionDate < "2023-12-31")

# Summary of results:
# Cambridge Bay:
# 2018 - 18299
# 2019 - 3477
# 2020 - 0
# 2021 - 128
# 2022 - 3223
# 2023 - 1188
# Kugluktuk:
# 2018 - 0
# 2019 - 4754
# 2020 - 0
# 2021 - 43
# 2022 - 6048
# 2023 - 0
# Gjoa Haven
# 2018 - 0
# 2019 - 0
# 2020 - 0
# 2021 - 3658
# 2022 - 6502
# 2023 - 0
# Kugaaruk
# 2018 - 0
# 2019 - 0
# 2020 - 0
# 2021 - 1013
# 2022 - 6011
# 2023 - 0
# Taloyoak
# 2018 - 0
# 2019 - 0
# 2020 - 0
# 2021 - 0
# 2022 - 4102
# 2023 - 0

# Great, now let's determine how many of the BINs are present in each community for more than one year
# For now, let's focus on Cambridge Bay and Kugluktuk
# How many distinct BINs are found in each community?
# How many of those persist year after year?

# Make a list of the distinct BINs for Cambridge Bay
cbay_2018_bins <- unique(cbay_2018$bin.uri)
cbay_2019_bins <- unique(cbay_2019$bin.uri)
cbay_2020_bins <- unique(cbay_2020$bin.uri)
cbay_2021_bins <- unique(cbay_2021$bin.uri)
cbay_2022_bins <- unique(cbay_2022$bin.uri)
cbay_2023_bins <- unique(cbay_2023$bin.uri)
# Great, we have hits for every year except 2020. Which makes sense, we didn't sample in 2020 due to COVID
# Let's create a data frame out of those
cbay_2018_bins <- data.frame(
  bin.uri = cbay_2018_bins
)
cbay_2019_bins <- data.frame(
  bin.uri = cbay_2019_bins
)
cbay_2020_bins <- data.frame(
  bin.uri = cbay_2020_bins
)
cbay_2021_bins <- data.frame(
  bin.uri = cbay_2021_bins
)
cbay_2022_bins <- data.frame(
  bin.uri = cbay_2022_bins
)
cbay_2023_bins <- data.frame(
  bin.uri = cbay_2023_bins
)

# Now, how many of the BINs are present in 2018 onwards?
test_2019 <- cbay_2018_bins %>%
  filter(bin.uri %in% cbay_2019_bins[,1])
# Interesting. There are 433 BINs in 2018 and 2019
test_2021 <- test_2019 %>%
  filter(bin.uri %in% cbay_2021_bins[,1])
# Even more interesting, 20 BINs are also in 2021
test_2022 <- test_2021 %>%
  filter(bin.uri %in% cbay_2022_bins[,1])
# And 15 of those were found in 2022 too!
test_2023 <- test_2022 %>%
  filter(bin.uri %in% cbay_2023_bins[,1])
# And 8 of those were found in 2023 too!

# Let's do something with that. Let's take those 8 BINs and see what they are.
cambridgebay_8bins <- cambridgebay %>%
  filter(bin.uri %in% c("BOLD:AAA6916","BOLD:AAB3041","BOLD:AAB7094","BOLD:AAB7975","BOLD:AAC5935","BOLD:AAF7753","BOLD:AAG5689"))
# Fascinating stuff. Those 8 BINs are made up of:
# 194 Araneae
# 127 Coleoptera
# 354 Diptera
# 144 Entomobryomorpha

